The JVM runs seems to be failing with:

```
2020-05-14T05:16:54.7680382Z [ERROR] Errors: 
2020-05-14T05:16:54.7681436Z [ERROR]   DevMojoIT.testThatApplicationRecoversStartupIssue:539->RunAndCheckMojoTestBase.runAndExpectError:86 Â» ConditionTimeout
```

--- a/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/samesite/SameSiteCookieTestCase.java
+++ b/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/samesite/SameSiteCookieTestCase.java
@@ -1,0 +1,50 @@
+package io.quarkus.vertx.http.samesite;
+
+import java.util.HashMap;
+import java.util.function.Supplier;
+
+import org.hamcrest.Matchers;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+import io.restassured.matcher.RestAssuredMatchers;
+
+public class SameSiteCookieTestCase {
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(new Supplier<JavaArchive>() {
+                @Override
+                public JavaArchive get() {
+                    return ShrinkWrap.create(JavaArchive.class)
+                            .addClasses(SetCookieHandler.class)
+                            .addAsResource(new StringAsset(
+                                    "quarkus.http.same-site-cookie.cookie1.value=Lax\n" +
+                                            "quarkus.http.same-site-cookie.cookie2.value=Lax\n" +
+                                            "quarkus.http.same-site-cookie.cookie2.case-sensitive=true\n" +
+                                            "quarkus.http.same-site-cookie.cookie3.value=None\n"),
+                                    "application.properties");
+                }
+            });
+
+    @Test
+    public void testSameSiteCookie() {
+        RestAssured.get("/cookie")
+                .then().cookies(new HashMap<>())
+                .cookie("cookie1", RestAssuredMatchers.detailedCookie().sameSite("Lax"))
+                .cookie("COOKIE2", RestAssuredMatchers.detailedCookie().sameSite(Matchers.nullValue()))
+                .cookie("cookie3", RestAssuredMatchers.detailedCookie().sameSite("None"))
+                .cookie("cookie3", RestAssuredMatchers.detailedCookie().secured(true));
+
+        RestAssured.with().header("user-agent", "Chromium/53 foo").get("/cookie")
+                .then().cookies(new HashMap<>())
+                .cookie("cookie1", RestAssuredMatchers.detailedCookie().sameSite("Lax"))
+                .cookie("COOKIE2", RestAssuredMatchers.detailedCookie().sameSite(Matchers.nullValue()))
+                .cookie("cookie3", RestAssuredMatchers.detailedCookie().sameSite(Matchers.nullValue()));
+    }
+}

--- a/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/samesite/SetCookieHandler.java
+++ b/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/samesite/SetCookieHandler.java
@@ -1,0 +1,27 @@
+package io.quarkus.vertx.http.samesite;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+
+import io.vertx.core.Handler;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.impl.CookieImpl;
+
+@ApplicationScoped
+public class SetCookieHandler {
+
+    public void handler(@Observes Router router) {
+        router.route("/cookie").handler(new Handler<RoutingContext>() {
+            @Override
+            public void handle(RoutingContext event) {
+                event.response().addCookie(new CookieImpl("cookie1", "value1"));
+                event.response().addCookie(new CookieImpl("COOKIE2", "VALUE2"));
+                event.response().addCookie(new CookieImpl("cookie3", "value3"));
+                event.response().addCookie(new CookieImpl("COOKIE4", "VALUE4"));
+                event.response().addCookie(new CookieImpl("foo", "foo"));
+                event.response().end();
+            }
+        });
+    }
+}

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpConfiguration.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpConfiguration.java
@@ -1,6 +1,7 @@
 package io.quarkus.vertx.http.runtime;
 
 import java.time.Duration;
+import java.util.Map;
 import java.util.Optional;
 import java.util.OptionalInt;
 
@@ -182,13 +183,19 @@ public class HttpConfiguration {
 
     /**
      * If this is true then the request start time will be recorded to enable logging of total request time.
-     * 
+     *
      * This has a small performance penalty, so is disabled by default.
      */
     @ConfigItem
     public boolean recordRequestStartTime;
 
     AccessLogConfig accessLog;
+
+    /**
+     * Configuration that allows setting the same site attributes for cookies.
+     */
+    @ConfigItem
+    public Map<String, SameSiteCookieConfig> sameSiteCookie;
 
     public int determinePort(LaunchMode launchMode) {
         return launchMode == LaunchMode.TEST ? testPort : port;

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/SameSiteCookieConfig.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/SameSiteCookieConfig.java
@@ -1,0 +1,40 @@
+package io.quarkus.vertx.http.runtime;
+
+import io.quarkus.runtime.annotations.ConfigGroup;
+import io.quarkus.runtime.annotations.ConfigItem;
+import io.vertx.core.http.CookieSameSite;
+
+/**
+ * Configuration that allows for automatically setting the SameSite attribute on cookies
+ *
+ * As some API's (Servlet, JAX-RS) don't current support this attribute this config allows
+ * it to be set based on the cookie name pattern.
+ */
+@ConfigGroup
+public class SameSiteCookieConfig {
+
+    /**
+     * If the cookie pattern is case sensitive
+     */
+    @ConfigItem
+    public boolean caseSensitive;
+
+    /**
+     * The value to set in the samesite attribute
+     */
+    @ConfigItem
+    public CookieSameSite value;
+
+    /**
+     * Some User Agents break when sent SameSite=None, this will detect them and avoid sending the value
+     */
+    @ConfigItem(defaultValue = "true")
+    public boolean enableClientChecker;
+
+    /**
+     * If this is true then the 'secure' attribute will automatically be sent on
+     * cookies with a SameSite attribute of None.
+     */
+    @ConfigItem(defaultValue = "true")
+    public boolean addSecureForNone;
+}

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/SameSiteNoneIncompatibleClientChecker.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/SameSiteNoneIncompatibleClientChecker.java
@@ -1,0 +1,138 @@
+package io.quarkus.vertx.http.runtime;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * A utility class that can check known user agents which are known to be incompatible with SameSite=None attribute.
+ * <p>
+ * <ul>
+ * <li>Versions of Chrome from Chrome 51 to Chrome 66 (inclusive on both ends).
+ * These Chrome versions will reject a cookie with `SameSite=None`. This also
+ * affects older versions of Chromium-derived browsers, as well as Android WebView.
+ * This behavior was correct according to the version of the cookie specification
+ * at that time, but with the addition of the new "None" value to the specification,
+ * this behavior has been updated in Chrome 67 and newer. (Prior to Chrome 51,
+ * the SameSite attribute was ignored entirely and all cookies were treated as if
+ * they were `SameSite=None`.)</li>
+ * <li>Versions of UC Browser on Android prior to version 12.13.2. Older versions
+ * will reject a cookie with `SameSite=None`. This behavior was correct according
+ * to the version of the cookie specification at that time, but with the addition of
+ * the new "None" value to the specification, this behavior has been updated in newer
+ * versions of UC Browser.
+ * <li>Versions of Safari and embedded browsers on MacOS 10.14 and all browsers on iOS 12.
+ * These versions will erroneously treat cookies marked with `SameSite=None` as if they
+ * were marked `SameSite=Strict`. This bug has been fixed on newer versions of iOS and MacOS.
+ * </ul>
+ * <p>
+ * 
+ * @see <a href="https://www.chromium.org/updates/same-site/incompatible-clients">SameSite=None: Known Incompatible Clients</a>.
+ */
+final class SameSiteNoneIncompatibleClientChecker {
+
+    /**
+     * User Agents Regex Patterns
+     */
+    private static final Pattern IOS_PATTERN = Pattern.compile("\\(iP.+; CPU .*OS (\\d+)[_\\d]*.*\\) AppleWebKit\\/");
+    private static final Pattern MACOSX_PATTERN = Pattern
+            .compile("\\(Macintosh;.*Mac OS X (\\d+)_(\\d+)[_\\d]*.*\\) AppleWebKit\\/");
+    private static final Pattern SAFARI_PATTERN = Pattern.compile("Version\\/.* Safari\\/");
+    private static final Pattern MAC_EMBEDDED_BROWSER_PATTERN = Pattern.compile(
+            "^Mozilla\\/[\\.\\d]+ \\(Macintosh;.*Mac OS X [_\\d]+\\) AppleWebKit\\/[\\.\\d]+ \\(KHTML, like Gecko\\)$");
+    private static final Pattern CHROMIUM_PATTERN = Pattern.compile("Chrom(e|ium)");
+    private static final Pattern CHROMIUM_VERSION_PATTERN = Pattern.compile("Chrom[^ \\/]+\\/(\\d+)[\\.\\d]* ");
+    // private static final Pattern UC_BROWSER_PATTERN = Pattern.compile("UCBrowser\\/");
+    private static final Pattern UC_BROWSER_VERSION_PATTERN = Pattern.compile("UCBrowser\\/(\\d+)\\.(\\d+)\\.(\\d+)[\\.\\d]* ");
+
+    public static boolean shouldSendSameSiteNone(String useragent) {
+        return !isSameSiteNoneIncompatible(useragent);
+    }
+
+    // browsers known to be incompatible.
+    public static boolean isSameSiteNoneIncompatible(String useragent) {
+        return hasWebKitSameSiteBug(useragent) ||
+                dropsUnrecognizedSameSiteCookies(useragent);
+    }
+
+    static boolean hasWebKitSameSiteBug(String useragent) {
+        return isIosVersion(12, useragent) ||
+                (isMacosxVersion(10, 14, useragent) &&
+                        (isSafari(useragent) || isMacEmbeddedBrowser(useragent)));
+    }
+
+    static boolean dropsUnrecognizedSameSiteCookies(String useragent) {
+        if (isUcBrowser(useragent)) {
+            return !isUcBrowserVersionAtLeast(12, 13, 2, useragent);
+        }
+        return isChromiumBased(useragent) &&
+                isChromiumVersionAtLeast(51, useragent) &&
+                !isChromiumVersionAtLeast(67, useragent);
+    }
+
+    // Regex parsing of User-Agent String. (See note above!)
+
+    static boolean isIosVersion(int major, String useragent) {
+        Matcher m = IOS_PATTERN.matcher(useragent);
+        if (m.find()) {
+            // Extract digits from first capturing group.
+            return String.valueOf(major).equals(m.group(1));
+        }
+        return false;
+    }
+
+    static boolean isMacosxVersion(int major, int minor, String useragent) {
+        Matcher m = MACOSX_PATTERN.matcher(useragent);
+        if (m.find()) {
+            // Extract digits from first and second capturing groups.
+            return String.valueOf(major).equals(m.group(1)) &&
+                    String.valueOf(minor).equals(m.group(2));
+        }
+        return false;
+    }
+
+    static boolean isSafari(String useragent) {
+        return SAFARI_PATTERN.matcher(useragent).find() &&
+                !isChromiumBased(useragent);
+    }
+
+    static boolean isMacEmbeddedBrowser(String useragent) {
+        return MAC_EMBEDDED_BROWSER_PATTERN.matcher(useragent).find();
+    }
+
+    static boolean isChromiumBased(String useragent) {
+        return CHROMIUM_PATTERN.matcher(useragent).find();
+    }
+
+    static boolean isChromiumVersionAtLeast(int major, String useragent) {
+        Matcher m = CHROMIUM_VERSION_PATTERN.matcher(useragent);
+        if (m.find()) {
+            // Extract digits from first capturing group.
+            int version = Integer.parseInt(m.group(1));
+            return version >= major;
+        }
+        return false;
+    }
+
+    static boolean isUcBrowser(String useragent) {
+        return useragent.contains("UCBrowser/");
+    }
+
+    static boolean isUcBrowserVersionAtLeast(int major, int minor, int build, String useragent) {
+        Matcher m = UC_BROWSER_VERSION_PATTERN.matcher(useragent);
+        if (m.find()) {
+            // Extract digits from three capturing groups.
+            int major_version = Integer.parseInt(m.group(1));
+            int minor_version = Integer.parseInt(m.group(2));
+            int build_version = Integer.parseInt(m.group(3));
+            if (major_version != major) {
+                return major_version > major;
+            }
+            if (minor_version != minor) {
+                return minor_version > minor;
+            }
+            return build_version >= build;
+        }
+        return false;
+    }
+
+}

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
@@ -6,19 +6,25 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.Map;
 import java.util.Optional;
+import java.util.TreeMap;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Executor;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.function.Supplier;
+import java.util.regex.Pattern;
 
 import javax.enterprise.event.Event;
 
@@ -62,6 +68,9 @@ import io.vertx.core.Handler;
 import io.vertx.core.Verticle;
 import io.vertx.core.Vertx;
 import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.Cookie;
+import io.vertx.core.http.CookieSameSite;
+import io.vertx.core.http.HttpHeaders;
 import io.vertx.core.http.HttpMethod;
 import io.vertx.core.http.HttpServer;
 import io.vertx.core.http.HttpServerOptions;
@@ -318,12 +327,19 @@ public class VertxHttpRecorder {
             quarkusWrapperNeeded = true;
         }
 
+        BiConsumer<Cookie, HttpServerRequest> cookieFunction = null;
+        if (!httpConfiguration.sameSiteCookie.isEmpty()) {
+            cookieFunction = processSameSiteConfig(httpConfiguration.sameSiteCookie);
+            quarkusWrapperNeeded = true;
+        }
+        BiConsumer<Cookie, HttpServerRequest> cookieConsumer = cookieFunction;
+
         if (quarkusWrapperNeeded) {
             Handler<HttpServerRequest> old = root;
             root = new Handler<HttpServerRequest>() {
                 @Override
                 public void handle(HttpServerRequest event) {
-                    old.handle(new QuarkusRequestWrapper(event));
+                    old.handle(new QuarkusRequestWrapper(event, cookieConsumer));
                 }
             };
         }
@@ -972,4 +988,56 @@ public class VertxHttpRecorder {
 
     private static final List<HttpMethod> CAN_HAVE_BODY = Arrays.asList(HttpMethod.POST, HttpMethod.PUT, HttpMethod.PATCH,
             HttpMethod.DELETE);
+
+    private BiConsumer<Cookie, HttpServerRequest> processSameSiteConfig(Map<String, SameSiteCookieConfig> httpConfiguration) {
+
+        List<BiFunction<Cookie, HttpServerRequest, Boolean>> functions = new ArrayList<>();
+        BiFunction<Cookie, HttpServerRequest, Boolean> last = null;
+
+        for (Map.Entry<String, SameSiteCookieConfig> entry : new TreeMap<>(httpConfiguration).entrySet()) {
+            Pattern p = Pattern.compile(entry.getKey(), entry.getValue().caseSensitive ? 0 : Pattern.CASE_INSENSITIVE);
+            BiFunction<Cookie, HttpServerRequest, Boolean> biFunction = new BiFunction<Cookie, HttpServerRequest, Boolean>() {
+                @Override
+                public Boolean apply(Cookie cookie, HttpServerRequest request) {
+                    if (p.matcher(cookie.getName()).matches()) {
+                        if (entry.getValue().value == CookieSameSite.NONE) {
+                            if (entry.getValue().enableClientChecker) {
+                                String userAgent = request.getHeader(HttpHeaders.USER_AGENT);
+                                if (userAgent != null
+                                        && SameSiteNoneIncompatibleClientChecker.isSameSiteNoneIncompatible(userAgent)) {
+                                    return false;
+                                }
+                            }
+                            if (entry.getValue().addSecureForNone) {
+                                cookie.setSecure(true);
+                            }
+                        }
+                        cookie.setSameSite(entry.getValue().value);
+                        return true;
+                    }
+                    return false;
+                }
+            };
+            if (entry.getKey().equals(".*")) {
+                //bit of a hack to make sure the pattern .* is evaluated last
+                last = biFunction;
+            } else {
+                functions.add(biFunction);
+            }
+        }
+        if (last != null) {
+            functions.add(last);
+        }
+
+        return new BiConsumer<Cookie, HttpServerRequest>() {
+            @Override
+            public void accept(Cookie cookie, HttpServerRequest request) {
+                for (BiFunction<Cookie, HttpServerRequest, Boolean> i : functions) {
+                    if (i.apply(cookie, request)) {
+                        return;
+                    }
+                }
+            }
+        };
+    }
 }

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/filters/QuarkusRequestWrapper.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/filters/QuarkusRequestWrapper.java
@@ -3,6 +3,7 @@ package io.quarkus.vertx.http.runtime.filters;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.function.BiConsumer;
 
 import org.jboss.logging.Logger;
 
@@ -31,10 +32,12 @@ public class QuarkusRequestWrapper extends AbstractRequestWrapper {
     private final AbstractResponseWrapper response;
 
     private final List<Handler<Void>> requestDoneHandlers = new ArrayList<>();
+    private final BiConsumer<Cookie, HttpServerRequest> cookieConsumer;
 
-    public QuarkusRequestWrapper(HttpServerRequest event) {
+    public QuarkusRequestWrapper(HttpServerRequest event, BiConsumer<Cookie, HttpServerRequest> cookieConsumer) {
         super(event);
-        this.response = new ResponseWrapper(delegate.response());
+        this.cookieConsumer = cookieConsumer;
+        this.response = new ResponseWrapper(delegate.response(), event);
         event.exceptionHandler(new Handler<Throwable>() {
             @Override
             public void handle(Throwable event) {
@@ -88,12 +91,14 @@ public class QuarkusRequestWrapper extends AbstractRequestWrapper {
 
     class ResponseWrapper extends AbstractResponseWrapper {
 
+        final HttpServerRequest request;
         Handler<Void> endHandler;
         Handler<Void> closeHandler;
         Handler<Throwable> exceptionHandler;
 
-        ResponseWrapper(HttpServerResponse delegate) {
+        ResponseWrapper(HttpServerResponse delegate, HttpServerRequest request) {
             super(delegate);
+            this.request = request;
             delegate.closeHandler(new Handler<Void>() {
                 @Override
                 public void handle(Void event) {
@@ -140,6 +145,14 @@ public class QuarkusRequestWrapper extends AbstractRequestWrapper {
         public HttpServerResponse endHandler(Handler<Void> handler) {
             this.endHandler = handler;
             return this;
+        }
+
+        @Override
+        public HttpServerResponse addCookie(Cookie cookie) {
+            if (cookieConsumer != null) {
+                cookieConsumer.accept(cookie, request);
+            }
+            return super.addCookie(cookie);
         }
     }
 

